topology/probes/k8s/linker.go

```diff
@@ -0,0 +1,133 @@
+/*
+ * Copyright 2018 IBM Corp.
+ *
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ *
+ */
+
+package k8s
+
+import (
+	"github.com/skydive-project/skydive/graffiti/graph"
+	"github.com/skydive-project/skydive/logging"
+	"github.com/skydive-project/skydive/probe"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// AreLinked return true if (a, b) should be linked
+type AreLinked func(a, b interface{}) bool
+
+// ABLinker basis for a simple A to B linker
+type ABLinker struct {
+	Manager   string
+	Type      string
+	Graph     *graph.Graph
+	ACache    *ResourceCache
+	BCache    *ResourceCache
+	AreLinked AreLinked
+}
+
+// NewEdgeIfMissing create a new edge
+func NewEdgeIfMissing(g *graph.Graph, i graph.Identifier, parent, child *graph.Node, m graph.Metadata) *graph.Edge {
+	if g.GetEdge(i) != nil {
```

Comment:
The test of the existence of the line is done by ResourceLinker

Reply:
If I remove this check then g.NewEdge() fails and I incorrectly log that as an error (because I can't distinguish between a "real" failure to a failure due to edge already existing).

Reply:
You are supposed to use `CreateEdge`, not `NewEdge` as it adds to the graph. AddEdge will be called by the linker

Reply:
ok - fixed